R/variance_ratio.R

```diff
@@ -33,100 +33,162 @@
 #'  Ecology 95, no. 6: 1693-1700. doi: 10.1890/13-0895.1
-#'  
-#'  Schluter, Dolph. (1984) "A Variance Test for Detecting Species Associations, 
+#'
+#'  Schluter, Dolph. (1984) "A Variance Test for Detecting Species Associations,
 #'  with Some Example Applications." Ecology 65, no. 3: 998-1005. doi:10.2307/1938071.
 #' @export
-#' @examples 
+#' @examples
 #'  data(knz_001d)
-#'  
+#'
 #'  # Calculate the variance ratio and CIs averaged within replicates
-#'  # Here the null model is repeated once, for final use it is recommended to set a 
+#'  # Here the null model is repeated once, for final use it is recommended to set a
 #'  # large bootnumber (eg, 10000)
-#'  
-#'  res_averagedreplicates <- variance_ratio(knz_001d, time.var = "year", species.var = "species", 
+#'
+#'  res_averagedreplicates <- variance_ratio(knz_001d, time.var = "year", species.var = "species",
 #'  abundance.var = "abundance", bootnumber = 1, replicate = "subplot")
-#'  
+#'
 #'  #Calculate the variance ratio and CIs for each replicate
-#'  
-#'  res_withinreplicates <- variance_ratio(knz_001d, time.var = "year", species.var = "species", 
+#'
+#'  res_withinreplicates <- variance_ratio(knz_001d, time.var = "year", species.var = "species",
 #'  abundance.var = "abundance", bootnumber = 1, replicate = "subplot", average.replicates = FALSE)
-variance_ratio <- function(df, time.var, species.var, abundance.var, bootnumber, replicate.var=NA,
-                        li=0.025, ui=0.975,  average.replicates=TRUE) {
-  
+variance_ratio <- function(df, time.var, species.var, abundance.var,
+                           bootnumber, replicate.var,
+                           average.replicates = TRUE, level = 0.95, ...) {
+
   # check to make sure abundance is numeric data
   check_numeric(df, time.var, abundance.var)
-   
+
   # if no replicates, calculate single variance ratio
-   if(is.na(replicate.var)) {
-        check_single_onerep(df, time.var, species.var)
-        VR <- variance_ratio_longformdata(df, time.var, species.var, abundance.var)
+  if  (missing(replicate.var)) {
+
+    check_single_onerep(df, time.var, species.var)
+    VR <- variance_ratio_longformdata(df, time.var, species.var, abundance.var)
+
+    ## null models
+
+    nullval <- cyclic_shift(df, time.var = time.var,
+                            species.var = species.var,
+                            abundance.var = abundance.var,
+                            method = variance_ratio_matrixdata,
+                            bootnumber = bootnumber)
+
+    nullout <- confint(nullval)
+
+    # bind actual value and CI
+    output <- cbind(nullout, VR)
+
+  } else {
+
+    # if multiple replicates, check all replicates have values
+    check_single(df, time.var, species.var, replicate.var)
+    ## if you give a replicate, it must be a factor. This gives users responsibility for order.
+    assertthat::assert_that(is.factor(df[[replicate.var]]))
+
```

Comment:
I disagree - I'd rather that it could be either a character or factor, and
generally I think it should be a character. We only coerce it a factor if
it was a factor and there were factor levels that weren't represented in
the dataset. I did this because our big test dataset had this scenario a
lot, I'd hope that most datasets dont.

On Fri, Mar 18, 2016 at 1:04 AM, Andrew MacDonald <notifications@github.com>
wrote:

> In R/variance_ratio.R
> <https://github.com/laurenmh/codyn/pull/67#discussion_r56619352>:
>
> > +                            abundance.var = abundance.var,
> > +                            method = variance_ratio_matrixdata,
> > +                            bootnumber = bootnumber)
> > +
> > +    nullout <- confint(nullval)
> > +
> > +    # bind actual value and CI
> > +    output <- cbind(nullout, VR)
> > +
> > +  } else {
> > +
> > +    # if multiple replicates, check all replicates have values
> > +    check_single(df, time.var, species.var, replicate.var)
> > +    ## if you give a replicate, it must be a factor. This gives users responsibility for order.
> > +    assertthat::assert_that(is.factor(df[[replicate.var]]))
> > +
>
> There used to be a line which checked replicate.var for being a factor,
> and if it was not (i think?) it coerced it. Here i'm suggesting we just
> force users to make it a factor themselves (safer, in case there is a
> particular order they want/expect in the output)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/laurenmh/codyn/pull/67/files/f3e38639882cb3e07b489b62de4bd6214853b215#r56619352>
>